### PR TITLE
Workaround for Holley DTZ541

### PR DESCRIPTION
**Pull request by hacki11**
- handling incorrect sml_time object
- support crc-16/kermit algorithm

### CODE REVIEW

File: sml/src/sml_time.c [hunk 1]
```diff
@@ -38,18 +38,29 @@ sml_time *sml_time_parse(sml_buffer *buf) {
...
-		goto error;
-	}
+    // workaround Holley DTZ541
+    // if SML_ListEntry valTime (SML_Time) is given there are missing bytes:
+    // 0x72: indicate a list for SML_Time with 2 entries
+    // 0x62 0x01: indicate secIndex
+    // instead, the DTZ541 starts with 0x65 + 4 bytes secIndex
+    // the workaround will add this information during parsing
+    if (sml_buf_get_current_byte(buf) == (SML_TYPE_UNSIGNED | 5)) {
+        u8 *tag = malloc(sizeof(u8));
```
Comment (mbehr1) [inline sml/src/sml_time.c]: this code has no effect. tag is a local variable that runs out of scope next line... (only impact: memory leak)
you might want to set
```
tme->tag = SML_TIME_SEC_INDEX;
```
?

Comment (hacki11) [reply sml/src/sml_time.c]: thought `tme->tag` is a pointer to a `u8`. Thats why i allocated the byte here.
`tme->tag` will be freed in `line 108: sml_number_free(tme->tag)`.

please correct me if iam wrong on that one

Comment (mbehr1) [reply sml/src/sml_time.c]: hmm. I didn't check the type from tml->tag. But in any case you need to assign it then e.g.
```
tme->tag = malloc(...) or
tme->tag = tag;
```

Comment (hacki11) [reply sml/src/sml_time.c]: ah, now i can see it :D 
good point!

File: sml/src/sml_time.c [hunk 2]
```diff
@@ -38,18 +38,29 @@ sml_time *sml_time_parse(sml_buffer *buf) {
...
-	}
+    // workaround Holley DTZ541
+    // if SML_ListEntry valTime (SML_Time) is given there are missing bytes:
+    // 0x72: indicate a list for SML_Time with 2 entries
+    // 0x62 0x01: indicate secIndex
+    // instead, the DTZ541 starts with 0x65 + 4 bytes secIndex
+    // the workaround will add this information during parsing
+    if (sml_buf_get_current_byte(buf) == (SML_TYPE_UNSIGNED | 5)) {
+        tme->tag = malloc(sizeof(u8));
+        *tme->tag = SML_TIME_SEC_INDEX;
```
Comment (mbehr1) [inline sml/src/sml_time.c]: reads dangerous for me. I'd prefer *(tme->tag) = ... but according to c operator precedence it's ok as is.

**Comment by hacki11** [comment]
Fixed review findings

**Comment by hacki11** [comment]
If there aren’t any more findings a merge would be nice

**Comment by andig** [comment]
Out of curiosity- why the unrelated whitespace change? Does that make anything more consistent?

**Comment by hacki11** [comment]
Seems that was my auto formatting of clion. Perhaps tab replacement with whitespaces and indentation fixes. Can revert that changes if you want.

**Comment by andig** [comment]
Please revert unwanted changes and check style- this repo is typically using tabs, you have spaces apparently. Reformatting the entire codebase would be a different issue.

**Comment by hacki11** [comment]
fixed

File: sml/src/sml_time.c [hunk 3]
```diff
@@ -38,18 +38,29 @@ sml_time *sml_time_parse(sml_buffer *buf) {
...
+    } else {
+        if (sml_buf_get_next_type(buf) != SML_TYPE_LIST) {
+            buf->error = 1;
+            goto error;
+        }
 
-	if (sml_buf_get_next_length(buf) != 2) {
-		buf->error = 1;
-		goto error;
-	}
```
Comment (andig) [inline sml/src/sml_time.c]: Hier stimmt noch die Einrückung nicht?

Comment (hacki11) [reply sml/src/sml_time.c]: Doch, der Block gehört nun in den else Zweig

**Comment by andig** [comment]
Fixes #75

**Comment by r00t-** [comment]
@hacki11: can you test https://github.com/volkszaehler/libsml/pull/96 ?

**Comment by hacki11** [comment]
I only use the lib and this is unit tested. But here are many people who tested the the software successfully 
https://www.photovoltaikforum.com/thread/134055-lesekopf-holley-dtz-541-zeba/?postID=2085184#post2085184

**Pull request merged**